### PR TITLE
Two-factor UI config and status display

### PR DESCRIPTION
**Pull request by rhinoau**
This PR improves visibility of 2FA status across users, adds config UI and defaults, and updates doco.

Inspired by my observation of the admin experience and doco out of this forum post: https://community.librenms.org/t/too-many-two-factor-failures/16323

Changes:
1. Set config defaults for twofactor and twofactor_lock
2. Add UI config for 2FA throttle-time and help text/descriptions
3. Add Manage Users grid 2FA column and visual indicator when twofactor is globally enabled
4. Documentation to suit and general cleanup/corrections

image
image

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: app/Http/Controllers/UserController.php [hunk 1]
```diff
@@ -56,6 +56,7 @@ public function index()
         return view('user.index', [
             'users' => User::orderBy('username')->get(),
             'multiauth' => User::query()->distinct('auth_type')->count('auth_type') > 1,
+            'twofactor' => UserPref::query()->where('pref', 'twofactor')->get('user_id'),
```
Comment (Jellyfrog) [inline app/Http/Controllers/UserController.php:59]: should use keyBy('user_id') instead so you dont need to loop it further down

File: resources/views/user/index.blade.php [hunk 2]
```diff
@@ -22,6 +22,9 @@
                     @if(\LibreNMS\Authentication\LegacyAuth::getType() == 'mysql')
                     <th data-column-id="enabled" data-formatter="enabled">@lang('Enabled')</th>
                     @endif
+                    @if(\LibreNMS\Config::get('twofactor'))
```
Comment (murrant) [inline resources/views/user/index.blade.php:25]: You can just use `@config('twofactor')`  and `@endconfig` (They are custom directives LibreNMS has)

File: resources/views/user/index.blade.php [hunk 3]
```diff
@@ -38,6 +41,11 @@
                             @if(\LibreNMS\Authentication\LegacyAuth::getType() == 'mysql')
                             <td>{{ $user->enabled }}</td>
                             @endif
+                            @if(\LibreNMS\Config::get('twofactor'))
+                                @if(isset($twofactor[$user->user_id]))
+                                <td>{{ $twofactor[$user->user_id]->pref }}</td>
```
Comment (murrant) [inline resources/views/user/index.blade.php:46]: I suggest you use `UserPref::getPref($user, 'twofactor')`, but I think the code should be updated to be more efficient (as below).  You should also pre-load the preferences relationship.  As in my other comment.

```php
    public static function getPref(User $user, $pref)
    {
        if ($user->relationLoaded('preferences')) {
            return optional($user->preferences->firstWhere('pref', $pref))->value;
        }

        return $user->preferences()->where('pref', $pref)->value('value');
    }
```

File: app/Http/Controllers/UserController.php [hunk 4]
```diff
@@ -56,6 +56,7 @@ public function index()
         return view('user.index', [
             'users' => User::orderBy('username')->get(),
```
Comment (murrant) [inline app/Http/Controllers/UserController.php:57]: Update this to:
`'users' => User::with('preferences')->orderBy('username')->get(),`

and delete the 'twofactor' line below.

File: resources/views/user/index.blade.php [hunk 5]
```diff
@@ -62,6 +70,11 @@
                             return '<span class="fa fa-fw fa-close text-danger"></span>';
                         }
                     },
+                    twofactor: function (column, row) {
+                        if(row['twofactor'] == 'twofactor') {
```
Comment (murrant) [inline resources/views/user/index.blade.php:74]: This will probably need a tweak after my sql changes.

**Review by Jellyfrog — APPROVED** [review]

File: resources/views/user/index.blade.php [hunk 6]
```diff
@@ -41,11 +41,11 @@
                             @if(\LibreNMS\Authentication\LegacyAuth::getType() == 'mysql')
                             <td>{{ $user->enabled }}</td>
                             @endif
-                            @if(\LibreNMS\Config::get('twofactor'))
-                                @if(isset($twofactor[$user->user_id]))
-                                <td>{{ $twofactor[$user->user_id]->pref }}</td>
+                            @config('twofactor')
+                                @if(app('App\Models\UserPref')->getPref($user, 'twofactor'))
```
Comment (murrant) [inline resources/views/user/index.blade.php:45]: This calls the static function non-statically.

**Pull request merged**